tests/this_param/object.js

```diff
@@ -0,0 +1,100 @@
+type I = {
+  anotherMethod(): number;
+};
+
+type J = {
+  anotherStaticMethod(): number;
+};
+
+class A {
+  x: number;
+  static x: number;
+  method(this: this & I): number {
+    return this.anotherMethod();
+  }
+  static staticMethod(this: J & Class<this>): number {
+    return this.anotherStaticMethod();
+  }
+}
+A.x = 1
+
+var a = new A();
+var n1: number = a.method(); // NG
+var n2: number = A.staticMethod(); // NG
+
+class B extends A {
+  anotherMethod(): string {
+    // NG: Incompatible with the `this` constraint on `method`.
+    return "another method";
+  }
+  static anotherStaticMethod(): string {
+    // NG: Incompatible with the `this` constraint on `staticMethod`.
+    return "another static method";
+  }
+}
+
+var b = new B();
+var s1: string = b.anotherMethod();
+var n3: number = b.method(); // NG
+var s2: string = B.anotherStaticMethod();
+
+// False negative
+// Although this is no good, an unnecessary error seems to be generated:
+// "`J & Class<this>` intersection. This type is incompatible with `B`". An
+// appropriate error is generated for the case, this is unneeded. The problem
+// may be rooted in ObjT lookups doing a unify--is some bound filter missing?
+var s3: number = B.staticMethod(); // NG
+
+class C extends A {
+  anotherMethod() {
+    return 5;
+  }
+  static anotherStaticMethod() {
+    return 2;
+  }
+}
+
+var c = new C();
+var n3: number = c.method();
+
+// False negative
+// "`C`. This type is incompatible with `J & Class<this>`, some incompatible
+// instantiation of `this`". Smells like a `subst` problem.
+var n4: number = C.staticMethod(); // OK
```

Comment:
"some incompatible instantiation of `this`" just sounds like the catchall error message to my ear now.  False negative probably due to `ClassT` induced leakiness of speculative matching similar to the "commutativity" problem.